hack/generator/pkg/functions/original_gvk_function.go

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package functions
+
+import (
+	"github.com/dave/dst"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astbuilder"
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+)
+
+// OriginalGVKFunction implements a function to return the GVK for the type on which it is called
+// We put these on the *storage* versions of our resource types, giving us a way to obtain the right type of instance
+// when the reconciler is working with ARM.
+//
+// func (resource *SomeResource) OriginalGVK() scheme.GroupVersionKind {
+//     return scheme.GroupVersionKind{
+//         Group: GroupVersion.Group,
+//         Version: resource.Spec.OriginalVersion,
+//         Kind: "SomeResource",
+//     }
+// }
+//
+type OriginalGVKFunction struct {
+	idFactory astmodel.IdentifierFactory
+}
+
+// Ensure OriginalGVKFunction properly implements Function
+var _ astmodel.Function = &OriginalGVKFunction{}
+
+// NewOriginalGVKFunction creates a new OriginalGVKFunction
+func NewOriginalGVKFunction(idFactory astmodel.IdentifierFactory) *OriginalGVKFunction {
+	return &OriginalGVKFunction{
+		idFactory: idFactory,
+	}
+}
+
+// Name returns the name of this function, which is always OriginalGVK()
+func (o OriginalGVKFunction) Name() string {
```

Comment:
I think since you're returning a `*OriginalGVKFunction` and asserting `var _ astmodel.Function = &OriginalGVKFunction{}` that the receiver here should also be a ptr (this applies to all the functions here)

Reply:
Fixed.